File: BinarySearch.java

public class BinarySearch {
	// This binary search algorithm is iterative in its nature.
	private static int binarySearchFunc(int array[], int key) {
		int index = -1;
		int low = 0;
		int high = array.length - 1;
		int mid;
		while (high >= low) {
			mid = (low + high) / 2;
			if (array[mid] == key) {
				index = mid;
				return index;
			} else if (array[mid] > key) {
				high = mid - 1;
			} else {
				low = mid + 1;
			}
		}
		return index;

	}

	public static void main(String[] args) {
		int array[] = { 10, 20, 30, 40, 50, 60, 70, 80, 90 };
		int key = 60;
		int res = binarySearchFunc(array, key);

		if (res != -1) {
			System.out.println("Key is found at index: " + (res+1));
		} else {
			System.out.println("Key is NOT found!");
		}
	}

}

File: binary_search.py
""" Given a sorted array of integers, return the index of the given key. Return -1 if not found.
https://www.geeksforgeeks.org/binary-search/
"""

def recursive_binary_search(arr, low, high, key):
	if low>high:
		return -1

	else:
		mid = low + (high - low)/2

		if arr[mid] == key:
			return mid
		elif arr[mid]>key:
			return recursive_binary_search(arr, low, mid - 1, key)
		else:
			return recursive_binary_search(arr, mid + 1, high, key)


def iterative_binary_search(arr, low, high, key):
	while low<=high:
		mid = low + (high - low)/2

		if arr[mid] == key:
			return mid
		elif arr[mid]<key:
			low = mid + 1
		else:
			high = mid - 1

	return -1


if __name__ == '__main__':
	arr = [1, 2, 3, 4, 5, 6, 7, 9, 10] # 8 is missing
	low, high = 0, len(arr) - 1
	key = 10

	print recursive_binary_search(arr, low, high, key)

	print iterative_binary_search(arr, low, high, key)

